Improve console handling of sanitycheck and support regex matching

nashif opened the pull request:
This introduces the concept of harness into sanitycheck and add initial support
for the Console harness which can match output from tests and match it against
a defined regex in the testcase file.

nashif committed:
sanitycheck: add harness classes

Add 2 classes, one to handle the current TestCase scenario, and one more
for handling generic Console with regex matching.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sample: synchronization: support console harness

Evaluate console output of sample.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sample: external_lib: support console harness

Evaluate console output of sample.

Enable on other platforms, this should not be whitelisted.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sample: philosophers: support console harness

Evaluate console output of sample.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sample: cpp_synchronization: support console harness

Evaluate console output of sample.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: support harness_config keyword in schema

Add harness_config to schema with basic options for now to support
console harness.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sample: kernel_event_logger: cleanup headers

Remove unused and duplicated header inclusion.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sample: hello_world: support console harness

Evaluate console output of sample.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: philosophers: test more scenarios

Enable 2 additional documented scenarios:
- preempt only
- coop only

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: use console harness for some samples

The output of those samples can be parsed and verified by sanitycheck,
so lets use the console harness for this.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: native: do not kill process when test completes

We can now send a signal and exit gracefully. This is to maintain
consistency across other runners and samples that can run natively but
do not have a test hook or macro to kill the application in completion.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: event_logger: remove filter and support harness

Evaluate output from sample and record success/failure.

Enable on other platforms, this should not be whitelisted.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: mpu_stack_guard: remove filter and support harness

Evaluate console output of sample.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: mem_domain_apis: remove filter and support harness

Evaluate console output of sample.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: nfc_hello: remove filter and support harness

Evaluate console output of sample.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: native: write output to run.log

Native handler should write output to a file, this is not happening
right now.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
test: mbedtls: remove while loop

Remove useless while loop from test, it serves no purpose.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
native: ignore format-truncation warnings (REVERTME)

This is enabled by default in GCC 7, disable for now. Tracked in
issue #5732

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: exclude some platforms due to bug [REVERTME]

Do not build for some unsupported architecture and exclude ARM qemu
platform due to bug #5735.

STACK_ALIGN 0x8
MPU STACK GUARD Test
Canary Initial Value = 0xf0cacc1a threads 0x200010e0
Canary = 0x20000240    Test not passed.
***** BUS FAULT *****
  Executing thread ID (thread): 0x200010e0
  Faulting instruction address:  0x209c
  Imprecise data bus error
Fatal fault in thread 0x200010e0! Aborting.
***** HARD FAULT *****
  Fault escalation (see below)
***** BUS FAULT *****
  Executing thread ID (thread): 0x200010e0
  Faulting instruction address:  0x1f4
  Imprecise data bus error
Fatal fault in ISR! Spinning...

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

dbkinder reviewed (approved):
+1 for doc changes

nashif committed:
samples: event_logger: match more lines during sanitycheck

Have the test run for long by matching more lines in the output.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: hello_world: disbale asserts on single thread (REVERTME)

For some reason, single thread is stuck when asserts are enabled.

This is tracked in #5733.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
recheck

andrewboie commented on scripts/sanitycheck:
```diff
@@ -2288,7 +2328,7 @@ def generate_coverage(outdir, ignores):
                  coveragefile],
                 stdout=coveragelog)
-        subprocess.call(["genhtml", "-output-directory",
+        subprocess.call(["genhtml", "--legend", "-output-directory",
```
nit: this change could be in another patch, but I don't care that much.

andrewboie commented on scripts/sanitycheck:
```diff
@@ -339,14 +339,19 @@ class NativeHandler(Handler):
         self.set_state("running", {})
 
-    def output_reader(self, proc, harness):
+    def _output_reader(self, proc, harness):
```
just curious why this method was renamed.

andrewboie commented on samples/application_development/external_lib/sample.yaml:
```diff
@@ -2,6 +2,10 @@ sample:
   name: External Library
 tests:
   test:
-    platform_whitelist: qemu_x86
```
maybe a line in the commit message on why this whitelist was removed.

nashif replied:
done

andrewboie commented on samples/philosophers/sample.yaml:
```diff
@@ -3,4 +3,13 @@ sample:
+      type: multi_line
+      repeat: 10
+      regex:
+        - ".*STARVING.*"
```
do the regexes need to be in this order or does it pass if it encounters any of these strings in indeterminate order?

andrewboie commented on samples/philosophers/sample.yaml:
```diff
@@ -1,15 +1,26 @@
 sample:
   name: Dining Philosophers
```
We could extend this more (maybe in another PR). Philosophers through ifdefs lets you use several different kernel object types for synchronization, i think we currently only test with mutexes IIRC.

nashif replied:
yes, good idea

andrewboie commented on samples/mpu/mpu_stack_guard_test/sample.yaml:
```diff
@@ -4,6 +4,8 @@ tests:
   test:
     tags: apps
```
this particular sample really need to be converted into a proper test.
i think there's a bug open on it somewhere.
nothing to change in your patch here, just whining

nashif replied:
agree, will address in another PR

andrewboie commented on samples/subsys/logging/kernel_event_logger/sample.yaml:
```diff
@@ -4,6 +4,10 @@ sample:
   name: Kernel event logger Sample
 tests:
   test:
-    platform_whitelist: qemu_x86
```
maybe a note in the commit message on why we can remove the whitelist

nashif replied:
done

andrewboie commented on arch/posix/CMakeLists.txt:
```diff
@@ -1,6 +1,7 @@
 zephyr_cc_option_ifdef(CONFIG_LTO -flto)
 
 zephyr_compile_options(
+  -Wno-format-truncation
```
our policy for REVERTME patches is that we need to file a bug on it and put the bug ID in the commit message

aescolar-ot replied:
In which file/line is this warning appearing and with what gcc version?

nashif replied:
done

andrewboie replied:
It just says "Tracked in issue" without an ID

nashif replied:
hmm, it was at the beginning of the line and was treated as comment, fixed now.

andrewboie commented on samples/drivers/crypto/sample.yaml:
```diff
@@ -17,7 +17,8 @@ tests:
   test-micro:
-    tags: micro
+    #FIXME fails in qemu
+    platform_exclude: qemu_nios2
```
this patch needs to be a REVERTME with a bug filed on it, this could be a symptom of something serious (or not)

andrewboie commented on samples/mpu/mpu_stack_guard_test/sample.yaml:
```diff
@@ -5,7 +5,8 @@ tests:
     # We do not handle segfaults here
-    platform_exclude: native_posix
+    arch_exclude: posix riscv32 nios2
+    platform_exclude: qemu_cortex_m3
```
this patch needs to be a REVERTME with a bug filed on it, this could be a symptom of something serious (or not)

andrewboie commented on samples/hello_world/sample.yaml:
```diff
@@ -13,6 +13,6 @@ common:
   singlethread:
     extra_args: CONF_FILE=prj_single.conf
-    filter: not CONFIG_BT and not CONFIG_GPIO_SCH
+    filter: not CONFIG_BT and not CONFIG_GPIO_SCH and ASSERT == 0
```
does this work? shouldn't this be "not CONFIG_ASSERT"?

andrewboie replied:
also please file a bug on this and link to the commit message of this patch

nashif replied:
it does not work as not CONFIG_ASSERT because ASSERT is being added by sanitycheck using -DCONFIG_ASSERT. I want to change this and make sanitycheck add kconfig overlays instead of defining this via cmake

aescolar-ot replied:
Regarding sanity check adding kconfig overlays: Would this be sufficient https://github.com/zephyrproject-rtos/zephyr/issues/5723?

nashif committed:
samples: philosophers: test different configurations

The demo can be configured to use different object types for its
synchronization, so test all of them.

The demo can also be configured to work with static objects or dynamic
objects, byt default the demo uses dynamic objects, add a test for
static objects.

Also, the demo can be configured to work with threads of the same
priority or not, so enable both options for testing

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: crypto: remove DEBUG/ASSERT from prj.conf

Those are added by sanitycheck, no need to have them enabled in the
project by default.

CONFIG_DEBUG is causing issues on qemu_nios2, see #5743.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

aescolar-ot commented on scripts/sanity_chk/harness.py:
```diff
@@ -0,0 +1,62 @@
+    def handle(self, line):
+        if self.type == "one_line":
+            pattern = re.compile(self.regex[0])
+            if pattern.match(line):
```
As commented in IRC:
This will only match the regular expression if the line does not have any prefix or postfix, which for some targets may be the case (for example a tracing tool may add a time stamp to each line of a trace). It would be nice if there was a `.*` added at the beginning and end of the expression.
So the line above would become ```pattern = re.compile(".*" + self.regex[0] + ".*")``` and the for multi-line ```pattern = re.compile(".*" + r + ".*")```

Otherwise, it would be nice to have at some point a more generic Handler mechanism. So that we could easily define that a given board has a different handler. Then we could ensure that the stream that comes out of the handler into this harness would be already free of extra characters.

andrewboie replied:
would re.search() also accomplish this?

aescolar-ot replied:
I think you are right

aescolar-ot commented on scripts/sanitycheck:
```diff
@@ -321,58 +327,72 @@ class NativeHandler(Handler):
+                    out_state = "failed valgrind"
 
-        returncode = subprocess.call(["GCOV_PREFIX=" + self.outdir, "gcov", self.sourcedir, "-s", self.outdir], shell=True)
+        returncode = subprocess.call(["GCOV_PREFIX=" + self.outdir, "gcov", self.sourcedir, "-b", "-s", self.outdir], shell=True)
```
Just a note: Most sample apps, are configured without COVERAGE and therefore will not account for this yet.
I guess when/if https://github.com/zephyrproject-rtos/zephyr/issues/5723 is implemented, we could use it to override their configuration. So that when run in sanitycheck, all tests and samples run in the POSIX arch would be compiled with coverage, instead of depending on the project configuration.

nashif committed:
native_posix: enable coverage by default

To get coverage data from all samples and tests, enable this by default.
Right now we only get coverage data from tests, however many samples we
currently have can run natively and can generate more coverage and
testing.

This should be removed once we have tests provide better coverage.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: use re.search to match output

Account for cases where there is a prefix to the output where exact
matching does not work.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

aescolar reviewed (approved):
Apart from this last commit message, this PR LGTM.
